grails-app/jobs/org/pih/warehouse/jobs/RefreshInventorySnapshotAfterTransactionJob.groovy

```diff
@@ -0,0 +1,89 @@
+import org.quartz.DisallowConcurrentExecution
+import org.quartz.JobExecutionContext
+import org.quartz.JobExecutionException
+
+@DisallowConcurrentExecution
+class RefreshInventorySnapshotAfterTransactionJob {
+
+    def concurrent = false
+    def grailsApplication
+    def inventorySnapshotService
+
+    static triggers = {}
+
+    def execute(JobExecutionContext context) {
+
+        Boolean enabled = ConfigurationHolder.config.openboxes.jobs.refreshInventorySnapshotAfterTransactionJob.enabled
+        log.info("Refresh inventory snapshots with data (enabled=${enabled}): " + context.mergedJobDataMap)
+        if (enabled) {
+
+            def jobDataMap = context.jobDetail.jobDataMap
+            Integer retryCount = jobDataMap.containsKey("retryCount") ? jobDataMap.getIntFromString("retryCount"):0
+            Integer maxRetryAttempts = ConfigurationHolder.config.openboxes.jobs.refreshInventorySnapshotJob.maxRetryAttempts?:3
+            Boolean retryOnError = ConfigurationHolder.config.openboxes.jobs.refreshInventorySnapshotJob.retryOnError?:false
+            if (retryOnError) {
+                log.info "Retry count: ${retryCount} / ${maxRetryAttempts}"
+                if (retryOnError && retryCount >= maxRetryAttempts) {
+                    JobExecutionException e = new JobExecutionException("Retries exceeded")
+                    e.setUnscheduleAllTriggers(true)
+                    throw e
+                }
+            }
+
+            boolean forceRefresh = context.mergedJobDataMap.getBoolean("forceRefresh")
+            def startTime = System.currentTimeMillis()
+            def userId = context.mergedJobDataMap.get('user')
+            def startDate = context.mergedJobDataMap.get('startDate')
+            String locationId = context.mergedJobDataMap.get('location')
+            String transactionId = context.mergedJobDataMap.get('transaction')?.id
+            Transaction transaction = Transaction.get(transactionId)
+            Location location = Location.get(locationId)
+
+            try {
+                log.info ("Refresh inventory snapshot " + retryCount + " out of " + maxRetryAttempts)
+
+                List productIds = transaction?.transactionEntries?.collect { it?.inventoryItem?.product?.id }?.unique()
+
+                productIds.each { productId ->
+                    Product product = Product.get(productId)
+
+                    // I feel like with this check, the inventory snapshot not always are properly populated (at least during the inventory import).
+                    // No inventory snapshots saves, despite it should be saved (for example two imports with quantities changed right after each other).
+                    if (forceRefresh) {
+                        inventorySnapshotService.deleteInventorySnapshots(location, product)
+                    }
+
+                    // When I call this one (without any date passed) I am getting more transaction dates than I feel like I should.
```

Comment:
Yes, this was what you were supposed to override so that it only ran on one date. What you're calling will update the inventory snapshots for all dates given the product and location. At some point we might need this process to be run on multiple dates. For example, when we update a transaction in the past we'd need to update all inventory snapshots from that date up until midnight tonight. But in the case we're trying to solve for we just want to update the inventory snapshot for tomorrow (i.e. new Date()+1).

Reply:
Yes, I had at the beginning the new Date() + 1, but then I went back to the one without a date.

Reply:
I'll readd this.